### PR TITLE
HDWallet error fix

### LINKED ISSUE #1485: Any String or mnemonic can be used to create wallet with  HDWallet(seedPhrase, passphrase)
**Describe the bug**
`val wallet = HDWallet("aa", "")`
Run android sample demo provided by this repo, even use "aa" as seedPhrase, wallet still successfully created.

**To Reproduce**
Run sample demo in this repo, change seedPhrase from legal mnemonic to "aa", run again

**Expected behavior**
wallet should not be  created

**Screenshots**
pic
**Additional context**

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

BREAKING changes!  (details below)
Improved error handling in HDWallet, do not allow invalid mnemonic phrases. Fixes #1485.
(Continuation of #1505 )

Breaking changes: 
- `HDWallet` creation can fail/throw on invalid input, mut be handled.  C++: throws.  Swift: returns optional, can return nil.  Kotlin: throws.
- Rename TWHDWalletCreateWithData to TWHDWalletCreateWithEntropy for clarity.  Swift: parameter name is `entropy` instead of `data`.  Kotlin not affected.
- Deprecate TWHDWalletIsValid (HDWallet.isValid) in favor of TWMnemonicIsValid (Mnemonic.isValid).

## Testing instructions

Unit tests, esp. negative HDWallet creation tests.

## Types of changes

* Bug fix (non-breaking change which fixes an issue) 

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/Mnemonic.h [hunk 1]
```diff
@@ -12,6 +12,10 @@ namespace TW {
 
 /// BIP39 Mnemonic Sentence handling.
 class Mnemonic {
+public:
+    static constexpr size_t MaxWords = 24;
+    static constexpr size_t MinWords = 12;
```
Comment (vikmeup) [inline src/Mnemonic.h:17]: I think by the protocol you can do 3 words too?

Comment (vikmeup) [reply src/Mnemonic.h:17]: How would it work if you have 24 words + password/passphrase?

File: include/TrustWalletCore/TWHDWallet.h [hunk 2]
```diff
@@ -49,6 +52,10 @@ TWData *_Nonnull TWHDWalletSeed(struct TWHDWallet *_Nonnull wallet);
 TW_EXPORT_PROPERTY
 TWString *_Nonnull TWHDWalletMnemonic(struct TWHDWallet *_Nonnull wallet);
 
+// Wallet entropy
+TW_EXPORT_PROPERTY
+TWData *_Nonnull TWHDWalletEntropy(struct TWHDWallet *_Nonnull wallet);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWHDWallet.h:57]: let's change `TWHDWalletCreateWithData` to `TWHDWalletCreateWithEntropy` as well

File: src/HDWallet.cpp [hunk 3]
```diff
@@ -32,42 +33,55 @@ HDNode getMasterNode(const HDWallet& wallet, TWCurve curve);
 const char* curveName(TWCurve curve);
 } // namespace
 
-HDWallet::HDWallet(int strength, const std::string& passphrase)
-    : seed(), mnemonic(), passphrase(passphrase) {
+HDWallet::HDWallet(int strength, const std::string& password)
+    : password(password) {
```
Comment (hewigovens) [inline src/HDWallet.cpp:37]: not sure why change `passphrase` to `password` it's the term used in bip39 with `mnemonic`

File: src/HDWallet.cpp [hunk 4]
```diff
@@ -32,42 +33,55 @@ HDNode getMasterNode(const HDWallet& wallet, TWCurve curve);
...
-    Data entropyRaw(32 + 1);
-    auto entropyBits = mnemonic_to_bits(mnemonic.c_str(), entropyRaw.data());
+void HDWallet::updateSeedAndEntropy() {
+    // generate seed from mnemonic
+    mnemonic_to_seed(mnemonic.c_str(), password.c_str(), seed.data(), nullptr);
+
+    // generate entropy from mnemonic
+    Data entropyRaw((Mnemonic::MaxWords * 11) / 8);
+    auto entropyBytes = mnemonic_to_bits(mnemonic.c_str(), entropyRaw.data()) / 8;
+    assert(entropyBytes != 0); // mnemonic must be valid
```
Comment (hewigovens) [inline src/HDWallet.cpp:80]: this `assert` seems covered by next line

File: tests/Keystore/StoredKeyTests.cpp [hunk 5]
```diff
@@ -361,8 +361,8 @@ TEST(StoredKey, MissingAddress) {
     auto key = StoredKey::load(TESTS_ROOT + "/Keystore/Data/missing-address.json");
     key.fixAddresses(password);
 
-    EXPECT_EQ(key.account(TWCoinTypeEthereum, nullptr)->address, "0x04De84ec355BAe81b51cD53Fdc8AA30A61872C95");
-    EXPECT_EQ(key.account(coinTypeBc, nullptr)->address, "bc1qe938ncm8fhdqg27xmxd7lq02jz9xh0x48r22lc");
+    EXPECT_EQ(key.account(TWCoinTypeEthereum, nullptr)->address, "0xA3Dcd899C0f3832DFDFed9479a9d828c6A4EB2A7");
```
Comment (hewigovens) [inline tests/Keystore/StoredKeyTests.cpp:364]: can you add your test phrases to comment? next time we won't have to change existing test data

File: tests/HDWalletTests.cpp [hunk 6]
```diff
@@ -19,6 +20,86 @@
 
 namespace TW {
 
+const auto mnemonic1 = "ripple scissors kick mammal hire column oak again sun offer wealth tomorrow wagon turn fatal";
+const auto password = "password";
+
+#define EXPECT_EXCEPTION(statement, exceptionMsg) \
```
Comment (hewigovens) [inline tests/HDWalletTests.cpp:26]: this looks useful, maybe we can put it in a standalone header file

**Review by hewigovens — COMMENTED** [review]
left some comments

File: tests/HDWalletTests.cpp (hunk 6 above)
Comment (catenocrypt) [reply tests/HDWalletTests.cpp]: Also planned to do it, I did it now.

File: tests/Keystore/StoredKeyTests.cpp (hunk 5 above)
Comment (catenocrypt) [reply tests/Keystore/StoredKeyTests.cpp:376]: The test data contained the test mnemonic with `back` as the last word, which is not a valid BIP39 word (worked before w/o check for valid mnemo).
I added now explicit check for the mnemonic in the loaded wallet.

File: src/HDWallet.cpp (hunk 3 above)
Comment (catenocrypt) [reply src/HDWallet.cpp:37]: My rationale was that recovery phrase and passphrase ('phrase' in both) is easy to confuse; moreover phrase is usually a sentence with multiple words and separator spaces, which is true for mnemonic (=recovery phrase) but generally not for the password.  So I think passphrase is more prone to confusion, password is more clear.
E.g. https://en.bitcoinwiki.org/wiki/Mnemonic_phrase uses 'password'.
But you are right, the official BIP39 uses 'passphrase'.
I will keep it as it is.

File: include/TrustWalletCore/TWHDWallet.h (hunk 2 above)
Comment (catenocrypt) [reply include/TrustWalletCore/TWHDWallet.h:57]: Changed, this is breaking change too.

**Review by hewigovens — APPROVED** [review]

File: src/Mnemonic.h (hunk 1 above)
Comment (catenocrypt) [reply src/Mnemonic.h:17]: These limits existed so far, I just made more explicit now.  Valid word counts are 12, 15, 18, 21, and 24.
3 words would be very insecure, easy to brute-force.
Passphrase is additional, used for storage encryption, it is not 25th word.

**Review by weishirongzhen — APPROVED** [review]

File: android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt [hunk 7]
```diff
@@ -22,16 +23,41 @@ class TestHDWallet {
...
+        assertEquals(Numeric.toHexString(hd.seed()), "0x354c22aedb9a37407adc61f657a6f00d10ed125efa360215f36c6919abd94d6dbc193a5f9c495e21ee74118661e327e84a5f5f11fa373ec33b80897d4697557d")
+    }
+
+    @Test
+    fun testCreateFromMnemonicInvalid() {
+        var exception = false
+        try {
+            val hd = HDWallet("THIS IS AN INVALID MNEMONIC", "")
+        } catch (ex: Exception) {
+            exception = true
```
Comment (vikmeup) [inline android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt:45]: Can you add assert for a specific error?

Comment (vikmeup) [reply android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt:45]: `AssertFail` or something similar

File: src/HDWallet.cpp [hunk 8]
```diff
@@ -62,12 +70,18 @@ HDWallet::~HDWallet() {
...
 }
 
-void HDWallet::updateEntropy() {
-    // generate entropy (from mnemonic)
-    Data entropyRaw(32 + 1);
-    auto entropyBits = mnemonic_to_bits(mnemonic.c_str(), entropyRaw.data());
+void HDWallet::updateSeedAndEntropy() {
+    // generate seed from mnemonic
+    // it is assumed that mnemonic is valid, enforced before calling
+    mnemonic_to_seed(mnemonic.c_str(), passphrase.c_str(), seed.data(), nullptr);
```
Comment (vikmeup) [inline src/HDWallet.cpp:76]: `mnemonic`, `passphrase`, `seed` should we make them private?

File: src/HDWallet.cpp [hunk 9]
```diff
@@ -62,12 +70,18 @@ HDWallet::~HDWallet() {
...
+
+    // generate entropy from mnemonic
+    Data entropyRaw((Mnemonic::MaxWords * 11) / 8);
```
Comment (vikmeup) [inline src/HDWallet.cpp:79]: what are these magic numbers?

File: src/HDWallet.cpp [hunk 10]
```diff
@@ -62,12 +70,18 @@ HDWallet::~HDWallet() {
...
+    auto entropyBytes = mnemonic_to_bits(mnemonic.c_str(), entropyRaw.data()) / 8;
+    assert(entropyBytes <= ((Mnemonic::MaxWords * 11) / 8) && entropyBytes >= ((Mnemonic::MinWords * 11) / 8));
```
Comment (vikmeup) [inline src/HDWallet.cpp:81]: where do you check that imported number of words <= MaxWords

Comment (catenocrypt) [reply src/HDWallet.cpp:81]: That's checked inside the c'tor with mnemonic, which calls Mnemonic::isValid(), which calls mnemonic_check().

https://github.com/trustwallet/wallet-core/blob/ar/hdwallet-error-fix/src/HDWallet.cpp#L49
https://github.com/trustwallet/wallet-core/blob/ar/hdwallet-error-fix/src/Mnemonic.cpp#L21
https://github.com/trustwallet/wallet-core/blob/ar/hdwallet-error-fix/trezor-crypto/crypto/bip39.c#L169

File: src/HDWallet.cpp (hunk 9 above)
Comment (catenocrypt) [reply src/HDWallet.cpp]: 11 is BitsPerWord, added now as constant.  8 is bits/byte.

File: src/HDWallet.cpp (hunk 8 above)
Comment (catenocrypt) [reply src/HDWallet.cpp:76]: It's safer that way (against coding mistakes).  Changed.
(General note: in most places in WC C++ code everything is public, which is more convenient, but more error prone).

**Comment by catenocrypt** [comment]
Review comments addressed

File: android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt [hunk 11]
```diff
@@ -38,13 +38,13 @@ class TestHDWallet {
...
-        var exception = false
+        var exception = ""
         try {
             val hd = HDWallet("THIS IS AN INVALID MNEMONIC", "")
         } catch (ex: Exception) {
-            exception = true
+            exception = ex::class.simpleName
         }
-        assertTrue(exception)
+        assertEquals(exception, "InvalidParameterException")
```
Comment (vikmeup) [inline android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt:47]: @vcoolish what's the better way to check an error?

Comment (vcoolish) [reply android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt:47]:
```kotlin
    @Test
    fun testCreateFromMnemonicInvalid() {
        try {
            HDWallet("THIS IS AN INVALID MNEMONIC", "")
            fail("Assert exception")
        } catch (ex: Exception) {
            assertTrue(ex is InvalidParameterException)
        }
    }
```

Comment (catenocrypt) [reply android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestHDWallet.kt:47]: I tried the more elegant assertFailWith (no try/catch needed), but it requires new dependency kotlin.test, which I don't want to add now
https://github.com/trustwallet/wallet-core/pull/1560
So I'll just stick with your suggestion.

**Review by vcoolish — APPROVED** [review]

**Review by vikmeup — APPROVED** [review]

**Pull request merged**